fpga_interchange/cell_parameters.cc

```diff
@@ -0,0 +1,319 @@
+/*
+ *  nextpnr -- Next Generation Place and Route
+ *
+ *  Copyright (C) 2021  Symbiflow Authors
+ *
+ *
+ *  Permission to use, copy, modify, and/or distribute this software for any
+ *  purpose with or without fee is hereby granted, provided that the above
+ *  copyright notice and this permission notice appear in all copies.
+ *
+ *  THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
+ *  WITH REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF
+ *  MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
+ *  ANY SPECIAL, DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
+ *  WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN
+ *  ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
+ *  OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
+ *
+ */
+
+#include "cell_parameters.h"
+
+#include <limits>
+
+#include "context.h"
+#include "nextpnr_assertions.h"
+#include "DeviceResources.capnp.h"
+#include "log.h"
+
+NEXTPNR_NAMESPACE_BEGIN
+
+CellParameters::CellParameters()
+    // 1'b0
+    : verilog_binary_re("([1-9][0-9]*)'b([01]+)$")
+    // 8'hF
+    , verilog_hex_re("([1-9][0-9]*)'h([0-9a-fA-F]+)$")
+    // 0b10
+    , c_binary_re("0b([01]+)$")
+    // 0xF
+    , c_hex_re("0x([0-9a-fA-F]+)$")
+    {}
+
+void CellParameters::init(const Context *ctx) {
+    for(const CellParameterPOD & cell_parameter : ctx->chip_info->cell_map->cell_parameters) {
+        IdString cell_type(cell_parameter.cell_type);
+        IdString parameter(cell_parameter.parameter);
+        auto result = parameters.emplace(std::make_pair(cell_type, parameter), &cell_parameter);
+        NPNR_ASSERT(result.second);
+    }
+}
+
+template<typename IntType>
+static DynamicBitarray<> to_bitarray(const IntType & value) {
```

Comment:
this seems like it would make more sense as a static member of `DynamicBitarray` (possibly the same for `parse_binary_bitstring` and `parse_hex_bitstring`)

Reply:
Sure

Reply:
Fixed